### PR TITLE
Reland #63639

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
This reverts commit 6536f65048ce910a5993375d489cf73e87caf765.

Prerequisite: https://github.com/flutter/flutter/issues/48679

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behavior with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by goderbauer** [comment]
(PR triage): What's the status of this PR? Is it waiting for #48679 to be fixed?

**Comment by LongCatIsLooong** [comment]
@goderbauer yes. Breaks internal tests otherwise.

**Comment by goderbauer** [comment]
https://github.com/flutter/flutter/issues/48679 has been fixed.

**Comment by goderbauer** [comment]
@LongCatIsLooong Any updates on this PR?

**Comment by LongCatIsLooong** [comment]
I'll see if I can make `RenderConstraintsTransformBox` an abstract class since it's difficult to implement `computeDryLayout` for it.

**Comment by skia-gold** [comment]
Gold has detected about 1 untriaged digest(s) on patchset 2.
View them at https://flutter-gold.skia.org/cl/github/64615

**Comment by skia-gold** [comment]
Gold has detected about 1 untriaged digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/64615

**Comment by skia-gold** [comment]
Gold has detected about 1 untriaged digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/64615

**Comment by skia-gold** [comment]
Gold has detected about 1 untriaged digest(s) on patchset 5.
View them at https://flutter-gold.skia.org/cl/github/64615

**Comment by LongCatIsLooong** [comment]
Ran TGP and there're 31 test failures. I'll split the PR into 2 parts and try to merge the safe part first.

File: packages/flutter/lib/src/rendering/shifted_box.dart [hunk 1]
```diff
@@ -797,6 +785,83 @@ class RenderUnconstrainedBox extends RenderAligningShiftedBox with DebugOverflow
...
+///
+///  * [RenderConstrainedBox], which renders a box which imposes constraints
+///    on its child.
+///  * [RenderConstrainedOverflowBox], which renders a box that imposes different
+///    constraints on its child than it gets from its parent, possibly allowing
+///    the child to overflow the parent.
+///  * [RenderSizedOverflowBox], a render object that is a specific size but
+///    passes its original constraints through to its child, which it allows to
+///    overflow.
+class RenderUnconstrainedBox extends RenderConstraintsTransformBox {
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/shifted_box.dart:812]: I wanted something like this when I was working on InteractiveViewer, and this is how I did it:

https://github.com/flutter/flutter/blob/c30fdfbe82d0d291c00ef199a35657990bba395c/packages/flutter/lib/src/widgets/interactive_viewer.dart#L1073-L1082

If I remember correctly it was actually really hard to do for sizes bigger than the screen.  Using a ConstrainedBox with infinite constraints didn't work, strangely.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

I left some comments that didn't end up as part of this review for some reason:
https://github.com/flutter/flutter/pull/77994#discussion_r595485047
https://github.com/flutter/flutter/pull/77994#discussion_r595484106

Just some stuff to think about.  It definitely took me awhile to understand exactly how this widget will behave in comparison to other widgets like ConstrainedBox.  It seems like that's due to the complexity of those widgets, not with this new one, though.

**Pull request closed without merging**